### ISSUE TITLE
blob to file conversion in formData

### ISSUE THREAD

**Issue by jimmywarting** [oven-sh/bun#4460]
### What version of Bun is running?

0.8.1

### What platform is your computer?

Darwin 22.4.0 arm64 arm

### What steps can reproduce the bug?

```js
var fd = new FormData()

var f1 = new File(['abc'], 'folder/readme.md', { type: 'text/plain' })
var f2 = new Blob(['abc'])

fd.set('f1', f1)
fd.set('f2', f2, 'customName.txt')

console.assert(fd.get('f1') instanceof File, 'adding a file remain as a file')
console.assert(fd.get('f2') instanceof File, 'adding a blob should be converted into a File')

console.assert(fd.get('f2').name === 'customName.txt', 'the name should be customName.txt')
console.assert(fd.get('f1').name === 'folder/readme.md', 'the name should be folder/readme.md')

new Response(fd).formData().then(fd => {
  console.assert(fd.get('f1') instanceof File, 'adding a file remain as a file')
  console.assert(fd.get('f2') instanceof File, 'adding a file remain as a file')

  console.assert(fd.get('f2').name === 'customName.txt', 'the name should be customName.txt')
  console.assert(fd.get('f1').name === 'folder/readme.md', 'the name should be folder/readme.md')
})
```

### What is the expected behavior?

not throwing any assertion errors

### What do you see instead?

- i got blobs back from `fd.get(x)`
- file names wasn't what i expected.

**Comment by AaronDewes** [comment]
The bug seems to be File to Blob conversion, not Blob to File. `f2` is working completely fine for me, but `f1` isn't, because `fd.set` expects a blob right now and ignores File properties. Not sure if it's related to this...

https://github.com/oven-sh/bun/blob/5c43744bce358f132a63d5002baffd7647d47a76/src/bun.js/bindings/JSDOMFile.cpp#L49-L50

**Comment by jimmywarting** [comment]
the assertion messages where not well, written (mostly copied) but my assertion test where correct, updated my example with better error message

**Comment by jimmywarting** [comment]
this is even worse than i tough: 

```js
console.log(new File([],'').constructor === Blob) // prints true, expect false
```

**Comment by scttcper** [comment]
Running into this as well on bun 1.0.2+37edd5a6e389265738e89265bcbdf2999cb81a49

```ts
const file = new File(['Hello'], 'file.txt', { type: 'text/plain' });
console.assert(file[Symbol.toStringTag] === 'File', `Expected 'File', got '${file[Symbol.toStringTag]}'`);
```

**Comment by jimmywarting** [comment]
In summary 
- Constructing a File should extend Blob and should be a instance of both File & Blob 
- The constructed File should have a filename and a symbol.toStringTag === File
- Appending a Blob to a FormData field should automatically upgrade it to a File `fd.set(x, blob); fd.get(x) !== blob`
- Appending a File to a FormData should not be cloned or changed, `fd.set(x, file); fd.get(x) === file`
- FormData<append | set> should also support a 3rd optional filename
- decoding FormData that contains file in a multipart should also return a File instance `fd = await response.formData()`
- it would be better if `Bun.file()` actually returned a File and not a Blob
  - the file should also include the files own lastModified timestamp
  - then i also wonder about the name, should it be full path, basename or configurable?

**Comment by jimmywarting** [comment]
And also something else i notice: 

Blob have stuff shoved to it that it should not have. namely: `Blob.prototype.name` and  `Blob.prototype.lastModified`
They should not exist on the `Blob`.
I discovered this error when i tried to extend (fix) the bad `File` with a basic (incomplete) File class
```js
class File extends Blob {
  constructor(fileBits, fileName, options = {}) {
    super(fileBits, options)

    this.lastModified = options.lastModified ?? Date.now()
    this.name = fileName
  }
}
```
b/c what i encountered was that i could not set `lastModified` and `name` to something that i wanted. I did get a
`Attempted to assign to readonly property` Error. 

So i had to add those to my own class:
```js
class File extends Blob {
  lastModified
  name
  
  constructor(fileBits, fileName, options = {}) {
    super(fileBits, options)

    this.lastModified = options.lastModified ?? Date.now()
    this.name = fileName
  }
}
```
which solved the issue... it's by no means a full-proof implementation of the File-constructor. i should probably have my own getters as well handle webidl conversion better

So...
- `Blob.prototype.name` and  `Blob.prototype.lastModified` have to be removed and added onto the `File` class instead where they belong.
- infact i think that every non-speced stuff should be moved onto the `BunFile` instead. that's to say: `exists`, `formData`, `json` and `writer` should be moved away from the Blob prototype.

**Comment by AaronDewes** [comment]
I've started working on a PR to fix this, I'll share it here soon. I just need to fix a "minor" bug: It segfaults while loading scripts that have `new File()`. In REPL, the bug is fixed by my patches though.

**Comment by jimmywarting** [comment]
`formData.get('file') === formData.get('file')` should evaluate to true, but dose not...

**Comment by Jarred-Sumner** [comment]
The issue at the top was fixed between Bun v1.1.10 and Bun v1.1.20

> 
formData.get('file') === formData.get('file') should evaluate to true, but dose not...

This does not evaluate to true in Chromium.

**Comment by jimmywarting** [comment]
> > formData.get('file') === formData.get('file') should evaluate to true, but dose not...
> 
> This does not evaluate to true in Chromium.

Chromium don't follow the spec...
When appending a Blob to a FormData, then it should be converted to a File immediately upon appending it.

https://issues.chromium.org/issues/40765469

you can also see that if you append a File, then `formData.get('file') === formData.get('file')` evaluate to true
